src/globus_sdk/scopes/builder.py

```diff
@@ -0,0 +1,161 @@
+
+        self._registered_scope_names: t.List[str] = []
+        self._register_scopes(known_scopes, self.urn_scope_string)
+        self._register_scopes(known_url_scopes, self.url_scope_string)
+
+    def _register_scopes(
+        self, scopes: ScopeBuilderScopes, transform_func: t.Callable[[str], str]
+    ) -> None:
+        scopes_dict = self._scopes_input_to_dict(scopes)
+        for scope_name, scope_val in scopes_dict.items():
+            self._registered_scope_names.append(scope_name)
+            setattr(self, scope_name, transform_func(scope_val))
+
+    def _scopes_input_to_dict(self, items: ScopeBuilderScopes) -> t.Dict[str, str]:
+        """
+        ScopeBuilders accepts many collection-style types of scopes. This function
+          normalizes all of those types into a standard {scope_name: scope_val} dict
+
+        Translation Map:
+        None => {}
+        "my-str" => {"my-str": "my-str"}
+        ["my-list"] => {"my-list": "my-list"}
+        ("my-tuple-key", "my-tuple-val") => {"my-tuple-key": "my-tuple-val"}
+        """
+        if items is None:
+            return {}
+        elif isinstance(items, str):
+            return {items: items}
+        elif isinstance(items, tuple):
+            return {items[0]: items[1]}
+        else:
+            items_dict = {}
+            for item in items:
+                if isinstance(item, str):
+                    items_dict[item] = item
+                else:
+                    items_dict[item[0]] = item[1]
+            return items_dict
+
+    @property
+    def scope_names(self) -> t.List[str]:
+        return self._classattr_scope_names + self._registered_scope_names
+
+    # custom __getattr__ instructs `mypy` that unknown attributes of a ScopeBuilder are
+    # of type `str`, allowing for dynamic attribute names
+    # to test, try creating a module with
+    #
+    #       from globus_sdk.scopes import TransferScopes
+    #       x = TransferScopes.all
+    #
+    # without this method, the assignment to `x` would fail type checking
+    # because `all` is unknown to mypy
+    #
+    # note that the implementation just raises AttributeError; this is okay because
+    # __getattr__ is only called as a last resort, when __getattribute__ has failed
+    # normal attribute access will not be disrupted
+    def __getattr__(self, name: str) -> str:
+        raise AttributeError(f"Unrecognized Attribute '{name}'")
+
+    def urn_scope_string(self, scope_name: str) -> str:
+        """
+        Return a complete string representing the scope with a given name for this
+        client, in the Globus Auth URN format.
+
+        Note that this module already provides many such scope strings for use with
+        Globus services.
+
+        **Examples**
+
+        >>> sb = ScopeBuilder("transfer.api.globus.org")
+        >>> sb.urn_scope_string("transfer.api.globus.org", "all")
+        "urn:globus:auth:scope:transfer.api.globus.org:all"
```

Comment:
Are these docstrings extracted by autodoc?

Reply:
Yes, they are. Do we need to `.. code-block: pycon` this? I thought that sphinx will read a leading `>>>` as the start of a console block, but maybe I'm not remembering the behavior correctly.

Reply:
Ah, yes, you're right. Cary on!